src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
+                nextByte = data[i];
+                if (nextByte != 'E' && nextByte != 'e')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedNextDigitEValueNotFound, nextByte);
+                }
+            }
+
+            Debug.Assert(nextByte == 'E' || nextByte == 'e');
+            i++;
+
+            signResult = ConsumeSign(ref data, ref i, nextByte);
+            if (signResult == ConsumeNumberResult.NeedMoreData)
+                return false;
+
+            Debug.Assert(signResult == ConsumeNumberResult.OperationIncomplete);
+
+            i++;
+            ConsumeNumberResult resultExponent = ConsumeIntegerDigits(ref data, ref i);
+            if (resultExponent == ConsumeNumberResult.NeedMoreData)
+                return false;
+            if (resultExponent == ConsumeNumberResult.Success)
+                goto Done;
+            if (resultExponent == ConsumeNumberResult.OperationIncomplete)
+            {
+                _lineBytePosition += i;
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndOfDigitNotFound, nextByte);
+            }
+
+        Done:
+            ValueSpan = data.Slice(0, i);
+            consumed = i;
+            return true;
+        }
+
+        private ConsumeNumberResult ConsumeNegativeSign(ref ReadOnlySpan<byte> data, ref int i)
+        {
+            byte nextByte = data[i];
+
+            if (nextByte == '-')
+            {
+                i++;
+                if (i >= data.Length)
+                {
+                    if (IsLastSpan)
+                    {
+                        _lineBytePosition += i;
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFoundEndOfData, nextByte);
+                    }
+                    else
+                        return ConsumeNumberResult.NeedMoreData;
+                }
+
+                nextByte = data[i];
+                if ((uint)(nextByte - '0') > '9' - '0')
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFound, nextByte);
+                }
+            }
+            return ConsumeNumberResult.OperationIncomplete;
+        }
+
+        private ConsumeNumberResult ConsumeZero(ref ReadOnlySpan<byte> data, ref int i)
+        {
+            i++;
+            byte nextByte = default;
+            if (i < data.Length)
+            {
+                nextByte = data[i];
+                if (JsonConstants.Delimiters.IndexOf(nextByte) >= 0)
+                    return ConsumeNumberResult.Success;
+            }
+            else
+            {
+                if (IsLastSpan)
+                    return ConsumeNumberResult.Success;
```

Comment:
Doesn't this mean `{ "number": 00` returns success?

Reply:
`ConsumeNumber` (which calls `TryGetNumber`) validates that we have some delimiter after the number. The reason why I don't do it here is because I also use `TryGetNumber` for single value json, where "0" is a valid json input and we shouldn't look for a delimiter.

I will add a test case for this regardless.